overlord/snapstate/snapstate.go

```diff
@@ -227,6 +230,124 @@ func Install(s *state.State, name, channel string, userID int, flags Flags) (*st
+func contains(ns []string, n string) bool {
+	i := sort.SearchStrings(ns, n)
+	if i >= len(ns) {
+		return false
+	}
+	return ns[i] == n
+}
+
+// RefreshCandidates gets a list of candidates for update
+// (call it with the state lock held)
+func RefreshCandidates(st *state.State, user *auth.UserState) ([]*snap.Info, error) {
+	updates, _, err := refreshCandidates(st, nil, user)
+	return updates, err
+}
+
+func refreshCandidates(st *state.State, names []string, user *auth.UserState) ([]*snap.Info, map[string]*SnapState, error) {
+	snapStates, err := All(st)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sort.Strings(names)
+
+	stateByID := make(map[string]*SnapState, len(snapStates))
+	candidatesInfo := make([]*store.RefreshCandidate, 0, len(snapStates))
+	for _, snapst := range snapStates {
+		if snapst.TryMode() || snapst.DevMode() {
+			// no automatic refreshes for trymode nor devmode
+			continue
+		}
+
+		// FIXME: snaps that are not active are skipped for now
+		//        until we know what we want to do
+		if !snapst.Active {
+			continue
+		}
+
+		snapInfo, err := snapst.CurrentInfo()
+		if err != nil {
+			// log something maybe?
+			continue
+		}
+
+		if len(names) > 0 && !contains(names, snapInfo.Name()) {
+			continue
+		}
+
+		stateByID[snapInfo.SnapID] = snapst
+
+		// get confinement preference from the snapstate
+		candidatesInfo = append(candidatesInfo, &store.RefreshCandidate{
+			// the desired channel (not info.Channel!)
+			Channel: snapst.Channel,
+			DevMode: snapst.DevModeAllowed(),
+			Block:   snapst.Block(),
+
+			SnapID:   snapInfo.SnapID,
+			Revision: snapInfo.Revision,
+			Epoch:    snapInfo.Epoch,
+		})
+	}
+
+	theStore := Store(st)
+
+	st.Unlock()
+	updates, err := theStore.ListRefresh(candidatesInfo, user)
+	st.Lock()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return updates, stateByID, nil
+}
+
+// UpdateMany updates everything from the given list of names that the
+// store says is updateable. If the list is empty, update everything.
+// (state must be locked)
+func UpdateMany(st *state.State, names []string, userID int) ([]string, []*state.TaskSet, error) {
+	user, err := userFromUserID(st, userID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	updates, stateByID, err := refreshCandidates(st, names, user)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	updated := make([]string, 0, len(updates))
+	tts := make([]*state.TaskSet, 0, len(updates))
+	for _, update := range updates {
+		snapst := stateByID[update.SnapID]
+		// XXX: this check goes away when update-to-local is done
+		if err := checkRevisionIsNew(update.Name(), snapst, update.Revision); err != nil {
+			continue
+		}
+
+		ss := &SnapSetup{
+			Channel:      snapst.Channel,
+			UserID:       userID,
+			Flags:        SnapSetupFlags(snapst.Flags),
+			DownloadInfo: &update.DownloadInfo,
+			SideInfo:     &update.SideInfo,
+		}
+
+		ts, err := doInstall(st, snapst, ss)
+		if err != nil {
+			// log?
```

Comment:
the only failure mode for doInstall seems to be if there's a conflict which seems ok to skip, otoh maybe having an explicit error type there and checking it explicitly and erroring otherwise would be saner long term